gnovm/pkg/gnolang/machine.go

```diff
@@ -237,6 +268,69 @@ func (m *Machine) RunMemPackage(memPkg *std.MemPackage, save bool) (*PackageNode
 	return pn, pv
 }
 
+func filterDuplicates(fset *FileSet) {
+	defined := make(map[Name]struct{}, 128)
+	for _, f := range fset.Files {
+		// returns true if name already defined
+		addName := func(n Name) bool {
+			if _, ok := defined[n]; ok {
+				return true // i--
+			}
+			defined[n] = struct{}{}
+			return false
+		}
+		for i := 0; i < len(f.Decls); i++ {
+			d := f.Decls[i]
+			var name Name
+			switch d := d.(type) {
+			case *FuncDecl:
+				name = d.Name
+				if d.IsMethod {
+					name = Name(derefStar(d.Recv.Type).String()) + "." + name
+				}
+			case *TypeDecl:
+				name = d.Name
+			case *ValueDecl:
+				// simpler, most common case
+				if len(d.NameExprs) == 1 {
+					name = d.NameExprs[0].Name
+					break
+				}
+
+				// more complex: filter out the NameExprs which are not
+				// redeclarations, and set them in d.NameExprs
+				newNames := make([]NameExpr, 0, len(d.NameExprs))
+				for _, nx := range d.NameExprs {
+					if !addName(nx.Name) {
+						newNames = append(newNames, nx)
```

Comment:
what if a name was previously defined but this new declaration changes the value?

case 1:
// orig file
var a = 1
// new file
var a, b, c = 2,3,4
// the outcome is that d.NamedExprs is [b,c]? isn't that strange?

case 2:
// orig file
var a = 1
// new file
var a = 2
// the outcome is that this declaration is ignored? isn't that strange too?

Reply:
Okay, so commits done for this:

- https://github.com/gnolang/gno/pull/859/commits/c64aaac6756a04de62e8df862b835f0a9eefbe53
- https://github.com/gnolang/gno/pull/859/commits/5592265c28f14e0b2fe3e6f6a18cddbf49dfc424 (bugfixing)


The caveat is that I had to move determining the type of a function to happen before we do `Define2`. Otherwise, `Define2` is always passed a zero-value FuncType. But since then the FuncType gets changed to be its actual value in `tryPredefine`, `Define2` panics because the two types are different (even though they are, at code-level, the same).